---
title: "ðŸ§¬ Differential Gene Expression Analysis"
output:
  html_document:
    code_folding: hide
    fig_height: 5
    fig_width: 10
params:
  path_config_json: ""
  count_data: NULL
  count_data_normalized: NULL
  out_path: ""
  save_intermediate: TRUE
  ensembl_version: 100
  biomart_attributes: !r NA
  minimum_count: 0
  minimum_padj: 0.05
  biotypes_filter: !r NA
---

This document contains many quality control plots of the differential expression (DE) analysis of your data.

Here is what you should do.

1. Check in **Overview & DESeq2 Run** if the sample ID's and the treatment are correct. Additionally, check if the comparisons are correct.
2. Use the plots in **Sample Similarity** if your experimental groups cluster together. If they don't, this might be a problem.
3. Check your differentially expressed genes

```{r setup, results = "hide", message = FALSE, warning = FALSE}

library(magrittr)
suppressWarnings(library(rmyknife))

# https://stackoverflow.com/questions/27992239/knitr-include-figures-in-report-and-output-figures-to-separate-files#comment71370705_27996644
knitr::opts_chunk$set(
  dev = c("png", "pdf"),
  fig.path = file.path(
    params$out_path,
    # Attach trailing '/'
    paste0("deseq_figures", .Platform$file.sep)
  )
)

ggplot2::theme_set(
  ggplot2::theme_minimal(
    base_size = 12
  )
)
ggplot2::theme_update(
  axis.ticks = ggplot2::element_line(color = "grey92"),
  axis.ticks.length = ggplot2::unit(.5, "lines"),
  panel.grid.minor = ggplot2::element_blank(),
  legend.title = ggplot2::element_text(size = 12),
  legend.text = ggplot2::element_text(color = "grey30"),
  plot.title = ggplot2::element_text(size = 18, face = "bold"),
  plot.subtitle = ggplot2::element_text(size = 12, color = "grey30"),
  plot.caption = ggplot2::element_text(size = 9, margin = ggplot2::margin(t = 15)),
  plot.title.position = 'plot'
)

rmd_params <- params

```

# {.tabset}

## Overview & DESeq2 Run

This analysis follows the [DESeq2 vigniette](https://bioconductor.org/packages/release/bioc/vignettes/DESeq2/inst/doc/DESeq2.html).

```{r print-parameters}

# Print settings as table
print_params <- params
print_params$count_data <- "not displayed"
print_params$count_data_normalized <- "not displayed"
rmyknife::print_params(print_params)

```

### Prepare DESeq2 Data Import

In this section we read in raw and normalized counts.
Additionally, we load the json file containing group assignments for each sample.

```{r load_data, message = FALSE}

settings <- jsonlite::read_json(rmd_params$path_config_json)

count_data <- rmd_params$count_data
count_data_normalized <- rmd_params$count_data_normalized

```

```{r prepare-col-data}

# Make empty tibble
col_data <- tibble::tibble()

# Iterate over group names and create treatment column
settings$groups %>%
  names() %>%
  purrr::walk(function(group_name) {
    coldata_group <- tibble::tibble(
      sample_name = settings$groups[group_name] %>% unlist(),
      treatment = group_name
    )
    # Attach rows to col_data tibble
    col_data <<- col_data %>%
      dplyr::bind_rows(coldata_group)
  })

# Next we get the required comparisons. Either they are specified with the JSON
# file or we automatically create them based on the treatment column
# Iterate over comparisons to create comparison list
comparisons <- list()
if (!is.null(settings$comparisons)) {
  comparisons <- settings$comparisons
} else {
  treatments <- col_data %>%
    dplyr::select("treatment") %>%
    dplyr::distinct() %>%
    dplyr::pull()
  for (i in 1:length(treatments)) {
    for (j in 1:length(treatments)) {
      if (i > j) {
        treatment_a <- treatments[i][[1]]
        treatment_b <- treatments[j][[1]]
        comparison_name <- paste0(
          treatment_a,
          "_vs_",
          treatment_b
        )
        comparisons[comparison_name][[1]] <- list(
          "group_a" = treatment_a,
          "group_b" = treatment_b
        )
      }
    }
  }
}

# The order in coldata must be the same as in count_data
# We achieve this by joining based on the id in count_data
col_data <- tibble::tibble(
  id = count_data %>%
    # Get column names
    colnames() %>%
    # Remove "gene_id" column which is the first one
    .[2:length(.)]
) %>%
  dplyr::left_join(
    col_data,
    by = c("id" = "sample_name")
  )

```

```{r check-tables}
# Check if we have samples in count table that we do not have a treatment for
undefined_samples <- col_data %>%
  dplyr::filter(
    is.na(treatment)
  )

# undefined_samples$id %>%
#   purrr::walk(function(na_sample) {
#     glue::glue("Removed '{na_sample}'' because it was assigned to no treatment group in the config") %>%
#       message()
#     count_data <<- count_data %>%
#       dplyr::select(-dplyr::contains(na_sample))
#     count_data_normalized <<- count_data_normalized %>%
#       dplyr::select(-dplyr::contains(na_sample))
#   })

count_data <-
  count_data %>%
  dplyr::select(undefined_samples$id)
count_data_normalized <-
  count_data_normalized %>%
  dplyr::select(-undefined_samples$id)

col_data <-
  col_data %>%
  dplyr::filter(!is.na(treatment)) %>%
  # Convert treatment variable to factor as required by DESeq2
  dplyr::mutate(treatment = as.factor(treatment))

# Convert counts to integer to comply with DEseq2 standards
count_data <-
  count_data %>%
  dplyr::mutate_if(is.double, round) %>%
  dplyr::mutate_if(is.double, as.integer)

```

```{r remove-biotypes, warning = FALSE, message = FALSE}

# Code from trapdiff
# Default report message.
remove_biotypes_report <- "We keep genes of all biotypes."
# Filter for biotypes of interest
if (!is.na(rmd_params$biotypes_filter)) {
  count_data <-
    count_data %>%
    rmyknife::attach_biomart(
      attributes = "gene_biotype",
      ensembl_id_var = "gene_id",
      ensembl_version = rmd_params$ensembl_version,
      verbose = FALSE
    ) %>%
    dplyr::filter(gene_biotype %in% rmd_params$biotypes_filter) %>%
    # Remove gene_biotype again
    dplyr::select(-gene_biotype)
  remove_biotypes_report <-
    glue::glue("We keep {count_data %>% nrow()} genes that are of bioype(s) {rmd_params$biotypes_filter}")
}

```

`r remove_biotypes_report`

```{r remove-mincount}

# Filter out genes where no treatment contains more than minimum_gene_count on average
entries_satisfying_mincount_and_biotypes <-
  count_data_normalized %>%
  # First, filter for the ones might already be filtered by biotype
  dplyr::filter(gene_id %in% count_data$gene_id) %>%
  tidyr::pivot_longer(-gene_id, values_to = "count_normalized") %>%
  # Attach treatment group
  dplyr::left_join(col_data, by = c("name" = "id")) %>%
  # Filter samples that are not associated with any treatments
  dplyr::filter(!is.na(treatment)) %>%
  dplyr::group_by(gene_id, treatment) %>%
  dplyr::summarise(mean_count_normalized = mean(count_normalized), .groups = "drop_last") %>%
  dplyr::filter(mean_count_normalized >= rmd_params$minimum_count) %>%
  dplyr::select(gene_id) %>%
  # Remove duplicates
  dplyr::distinct() %>%
  dplyr::pull()

count_data <- count_data %>%
  dplyr::filter(gene_id %in% entries_satisfying_mincount_and_biotypes)
```

We work with `r length(entries_satisfying_mincount_and_biotypes)` genes where at least one treatment has a mean gene normalized count expression of `r rmd_params$minimum_count` (this is minimum count and biotype filtering combined).

Here are the top entries of the raw and normalized count table:

```{r print-top-entries-raw}
count_data %>%
  head() %>%
  knitr::kable()

count_data_normalized %>%
  head() %>%
  knitr::kable()
```

The sample group assignment table is defined as follows:

```{r print-design-table}
col_data %>%
  knitr::kable()
```

Here are the comparisons we are running:

```{r print-comparisons}
comparisons %>% print()
```

### Create DESeq2 data and run differential expression analysis

```{r create-deseq2-data}
deseq_data <- DESeq2::DESeqDataSetFromMatrix(
  countData = count_data %>% as.data.frame(),
  colData = col_data,
  design = ~treatment,
  tidy = TRUE
)
```

```{r run-differential-expression-analysis}

deseq_result <- DESeq2::DESeq(deseq_data)

deseq_result_frame <- comparisons %>%
  purrr::imap(function(comparison, comparison_name) {
    paste0(
      "Get differential expression for comparison ",
      comparison_name,
      ", group_a: ",
      comparison$group_a,
      ", group_b: ",
      comparison$group_b
    ) %>%
      message()

    DESeq2::results(
      deseq_result,
      tidy = TRUE,
      # Define which group we want to extract
      contrast = c("treatment", comparison$group_a, comparison$group_b)
    ) %>%
      tibble::as_tibble() %>%
      # Keep the comparison as one variable in the data frame
      dplyr::mutate(comparison = comparison_name) %>%
      # Arrange by adjusted p-value
      dplyr::arrange(padj)
  }) %>%
  # Bind all data frames into one frame
  dplyr::bind_rows() %>%
  dplyr::rename(ensembl_gene_id = row)
```

Attach biomart parameters.

```{r attach-biomart, warning = FALSE, message = FALSE}
# For some reason caching this creates problems with the attached deseq_result frame
# To avoid this, we created a memoized version of the attach_biomart function
if (!is.na(rmd_params$biomart_attributes)) {
  deseq_result_frame <-
    deseq_result_frame %>%
    rmyknife::attach_biomart(
      ensembl_version = rmd_params$ensembl_version,
      ensembl_id_var = "ensembl_gene_id",
      attributes = rmd_params$biomart_attributes
    )
}
```

### Save results

With the differential expression analysis complete we can now store away the resulting files based on the input parameters.

```{r output-deseq-results}

out_path_results <- file.path(rmd_params$out_path, "deseq_diff")
out_path_intermediates <- file.path(out_path_results, "intermediates")


if (rmd_params$out_path != "") {
  glue::glue("You set the output file path to {out_path_results}") %>%  message()
  # Create the folder
  dir.create(out_path_results, recursive = TRUE)
  # Save csv
  csv_path <- file.path(out_path_results, "deseq2_diff.csv")
  glue::glue("Save CSV file to {csv_path}") %>% message()
  readr::write_csv(x = deseq_result_frame, file = csv_path)
  # Save Excel
  excel_path <- file.path(out_path_results, "deseq2_diff.xlsx")
  glue::glue("Save Excel file to {excel_path}") %>% message()
  rmyknife::write_xls(dat = deseq_result_frame, ExcelFileName = excel_path)
  # Save RDS
  if (rmd_params$save_intermediate) {
    dir.create(out_path_intermediates)
    # deseq2_diff
    deseq2_diff_path <- file.path(out_path_intermediates, "deseq2_diff.rds")
    glue::glue("Save deseq2_diff file to {deseq2_diff_path}") %>% message()
    saveRDS(object = deseq_result, file = deseq2_diff_path)
    # col_data
    col_data_path <- file.path(out_path_intermediates, "col_data.csv")
    glue::glue("Save col_data file to {col_data_path}") %>% message()
    readr::write_csv(x = col_data, file = col_data_path)
    # count_data
    count_data_path <- file.path(out_path_intermediates, "count_data.csv")
    glue::glue("Save count_data file to {count_data_path}") %>% message()
    readr::write_csv(x = count_data, file = count_data_path)
    # count_data_normalized
    count_data_normalized_path <- file.path(out_path_intermediates, "count_data_normalized.csv")
    glue::glue("Save count_data_normalized file to {count_data_normalized_path}") %>% message()
    readr::write_csv(x = count_data_normalized, file = count_data_normalized_path)
  }
}
```

## Sample Similarity

Here we look at the variance between the samples.

### Principal Component Analysis

The Principal Component Analysis (PCA) shows which samples are similar on a genomic level.
Ideally we want our experimental groups to cluster together.
If your samples do not cluster up in the PCA plot, there are going to be less differentially expressed genes, because the variance is too high.

We plot two versions of the PCA, one with and without labels.

<details>
<summary>PCA without labels</summary>

```{r pca}

pca_data <- DESeq2::plotPCA(
  DESeq2::vst(deseq_result, blind = FALSE),
  intgroup = "treatment",
  returnData = TRUE
)

pca_plot <- pca_data %>%
  ggplot2::ggplot(ggplot2::aes(x = PC1, y = PC2, label = name, colour = group)) +
  ggplot2::geom_point() +
  ggplot2::ggtitle("Principal Component Analysis") +
  ggplot2::xlab(paste0("Principal Component 1 (", round(attr(pca_data, "percentVar")[1] * 100), "% variance explained)")) +
  ggplot2::ylab(paste0("Principal Component 2 (", round(attr(pca_data, "percentVar")[2] * 100), "% variance explained)"))

pca_plot
```

</details>

```{r pca-with-labels}
pca_plot + ggrepel::geom_text_repel()
```

### Distance Matrix

This plot shows sample similarity.
It can help finding outliers and similar samples.
It captures the same information as the PCA, represented differently.
See the [DESeq2 manual](https://bioconductor.org/packages/release/bioc/vignettes/DESeq2/inst/doc/DESeq2.html#differential-expression-analysis) for more details.

```{r distance-matrix}

distance_matrix <- function(deseq_result) {
  vsdata <- DESeq2::vst(deseq_result, blind = FALSE)
  sampleDists <- dist(t(SummarizedExperiment::assay(vsdata)))
  sampleDistMatrix <- as.matrix(sampleDists)

  # rownames(sampleDistMatrix) <- paste(vsdata$condition, vsdata$type, sep="-")
  rownames(sampleDistMatrix) <- colnames(sampleDistMatrix)
  # colnames(sampleDistMatrix) <- NULL

  # colors <- colorRampPalette( rev(RColorBrewer::brewer.pal(9, "Blues")) )(255)
  pheatmap::pheatmap(sampleDistMatrix,
    clustering_distance_rows = sampleDists,
    clustering_distance_cols = sampleDists,
    col = colorRampPalette(RColorBrewer::brewer.pal(9, "Blues"))(255)
  )
}

distance_matrix(deseq_result)
```

## Differentially Expressed (DE) Genes

There are `r deseq_result_frame %>% tibble::as_tibble() %>% dplyr::filter(padj <= rmd_params$minimum_padj) %>% nrow()` entries differentially expressed.

*If you click on one of the download buttons above the table, give your browser a minute to process the request.*

### Data dictionary

| Column               | Description                                                                                                                                                                                               |
| -------------------- | --------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- |
| `ensembl_gene_id`    | [ensembl.org](https://www.ensembl.org/info/genome/stable_ids/index.html) gene stable ID                                                                                                                  |
| `baseMean`           | [DESeq2](https://doi.org/10.1186/s13059-014-0550-8) - average of the normalized count values, dividing by size factors (adjusting for different library size), taken over all samples.                    |
| `log2FoldChange`     | [DESeq2](https://doi.org/10.1186/s13059-014-0550-8) - effect size estimate between treatment and control. A log2 fold change of 1.5 means the gene expression is increased by 2^1.5 = 2.82.               |
| `lfcSE`              | [DESeq2](https://doi.org/10.1186/s13059-014-0550-8) - Uncertainty associated with `log2FoldChange`, representing the standard error.                                                                      |
| `stat`               | [DESeq2](https://doi.org/10.1186/s13059-014-0550-8) - `log2FoldChange`/`lfcSE`. This is used to calculate the `pvalue`.                                                                                   |
| `pvalue`             | [DESeq2](https://doi.org/10.1186/s13059-014-0550-8) - Probability that a fold change as strong as the observed one, or even stronger, would be seen under the situation described by the null hypothesis. |
| `padj`               | [DESeq2](https://doi.org/10.1186/s13059-014-0550-8) - Benjamini-Hochberg adjusted `pvalue`, adjustment for false discovery rate (FDR).                                                                    |
| `comparison`         | Pairwise comparison for the current DESeq2 output.                                                                                                                                                        |
| `external_gene_name` | [Ensembl Biomart](https://www.ensembl.org/info/data/biomart/index.html) gene name.                                                                                                                        |
| `gene_biotype`       | [Ensembl Biomart](https://www.ensembl.org/info/data/biomart/index.html) [Gencode gene biotype](https://www.gencodegenes.org/pages/biotypes.html).                                                         |
| `description`        | [Ensembl Biomart](https://www.ensembl.org/info/data/biomart/index.html) gene description.                                                                                                                 |

### DE genes table

Differentially expressed genes, all values rounded to 4 floating point decimals.

```{r print-de-entries}

deseq_result_frame %>%
  dplyr::filter(padj <= rmd_params$minimum_padj) %>%
  dplyr::mutate_if(is.double, function(x) {
    return(round(x, 4))
  }) %>%
  rmyknife::dt_datatable()
```

<details>
<summary>Unfiltered DESeq2 table output without rounding</summary>

```{r print-table-no-rounding}
rmyknife::dt_datatable(deseq_result_frame)
```

</details>


## DE Genes Plots

Let's look at some plots of the differentially expressed genes.

### Regulated Genes Heat Map

Heat map of top-regulated genes accross all samples. This outlines the sample similarity.

```{r heat-map-top-n-counts, warning = FALSE}

deseq_counts <- deseq_result %>%
  DESeq2::counts(
    normalized = TRUE,
    replaced = FALSE
  ) %>%
  as.data.frame() %>%
  tibble::rownames_to_column("ensembl_id") %>%
  tidyr::pivot_longer(-ensembl_id, "sample_name") %>%
  dplyr::rename(count = value) %>%
  # Attach adjusted p-value
  dplyr::left_join(
    .,
    deseq_result_frame %>%
      dplyr::select(ensembl_gene_id, padj, comparison),
    by = c("ensembl_id" = "ensembl_gene_id")
  ) %>%
  # Order by adjusted p-value
  dplyr::arrange(padj) %>%
  tibble::as_tibble()

# Iterate over all comparison names
deseq_counts %>%
  dplyr::select(comparison) %>%
  dplyr::distinct() %>%
  dplyr::pull() %>%
  purrr::walk(
    function(comparison_name) {
      # Draw the heat map based on the current comparison
      deseq_counts_tmp <- deseq_counts %>%
        dplyr::filter(comparison == comparison_name) %>%
        dplyr::filter(padj <= rmd_params$minimum_padj)
      if (deseq_counts_tmp %>% nrow() == 0) {
        return()
      }

      deseq_counts_tmp %>%
        # Create adjusted count column suitable for plotting
        dplyr::mutate(adj_count = log10(count + 1)) %>%
        # Heatmap.2
        dplyr::select(ensembl_id, sample_name, adj_count) %>%
        tidyr::pivot_wider(names_from = sample_name, values_from = adj_count) %>%
        tibble::column_to_rownames("ensembl_id") %>%
        as.matrix() %>%
        t() %>%
        (function(dat_heatmap) {
          # Check if data frame meets minimum requirement
          if (nrow(dat_heatmap) < 2 | ncol(dat_heatmap) < 2) {
            paste0(
              "Cannot print heat map for comparison ",
              comparison_name,
              ". It contains ",
              nrow(deseq_counts_tmp),
              " significant genes."
            ) %>%
              warning()
            return()
          }
          # Print heat map
          dat_heatmap %>% gplots::heatmap.2(
            dendrogram = "row",
            scale = "none",
            trace = "none",
            main = paste0(
              comparison_name,
              ". Top ",
              ncol(.),
              " genes"
            ),
            labCol = NA,
            # Set labels
            xlab = "Genes",
            col = colorRampPalette(RColorBrewer::brewer.pal(9, "Blues"))(200),
            key.title = expression(-log[10](count + 1)),
            margins = c(4, 15)
          )
        })
    }
  )
```

### Volcano Plot

A volcano plot maps fold change on the x axis against significance on the y-axis.
We create the plot for all comparisons available in the data.
The *top 10* differentially regulated genes are labeled.

```{r volcano-plot, fig.width = 12, fig.height = 6, results = "asis"}

deseq_result_frame %>%
  dplyr::select(comparison) %>%
  dplyr::distinct() %>%
  dplyr::pull() %>%
  purrr::walk(function(comparison_name) {
    # Get data frame that is already filtered for the current comparison
    deseq_result_frame_comparison <- deseq_result_frame %>%
      dplyr::filter(comparison == comparison_name)
    # Get table of significantly up- and down-regulated genes
    tibble::tibble(
      property = c(
        "Number of up-regulated genes",
        "Number of differentially expressed up-regulated genes",
        "Number of down-regulated genes",
        "Number of differentially expressed down-regulated genes"
      ),
      value = c(
        # Number of up-regulated genes
        deseq_result_frame_comparison %>%
          dplyr::filter(log2FoldChange > 0) %>%
          nrow(),
        # Number of differentially expressed up-regulated genes
        deseq_result_frame_comparison %>%
          dplyr::filter(log2FoldChange > 0 & padj <= rmd_params$minimum_padj) %>%
          nrow(),
        # Number of down-regulated genes
        deseq_result_frame_comparison %>%
          dplyr::filter(log2FoldChange < 0) %>%
          nrow(),
        # Number of differentially expressed down-regulated genes
        deseq_result_frame_comparison %>%
          dplyr::filter(log2FoldChange < 0 & padj <= rmd_params$minimum_padj) %>%
          nrow()
      )
    ) %>%
      knitr::kable() %>%
      print()
    # Create volcano plot
    volcano_plot <- deseq_result_frame_comparison %>%
      # Remove entries that cannot be drawn
      dplyr::filter(!is.na(padj) & !is.na(log2FoldChange)) %>%
      # Draw the plot
      ggplot2::ggplot(
        ggplot2::aes(
          x = log2FoldChange,
          y = -log10(padj),
          color = (padj <= rmd_params$minimum_padj) %>% ifelse(., "significant", "not significant")
        )
      ) +
      ggplot2::geom_point(
        alpha = 0.3,
        size = 0.5
      ) +
      ggplot2::scale_color_manual(values = c("grey", "blue")) +
      ggplot2::ggtitle(paste0("Comparison: ", comparison_name)) +
      ggplot2::xlab(expression(log[2](fc))) +
      ggplot2::ylab(expression(-log[10](adjusted ~ p ~ value))) +
      ggplot2::labs(colour = "Significance") +
      ggrepel::geom_text_repel(
        data = . %>%
          dplyr::arrange(padj) %>%
          head(10),
        mapping = ggplot2::aes(label = external_gene_name),
        size = 3
      ) +
      ggplot2::geom_hline(
        yintercept = -log10(rmd_params$minimum_padj),
        linetype = "dotted"
      )
    # Print volcano plot
    volcano_plot %>%
      print()
  })
```
