---
title: "Differential Gene Expression Analysis based on Salmon output of nf-core RNASeq pipeline."
output:
  html_notebook:
    toc: yes
    toc_float: yes
    code_folding: hide
params:
  path_config_json: ""
  path_salmon_counts: ""
  top_regulated_genes_for_plotting: 10
  dev: TRUE
---


```{r helper}
ren <- function() {
  rmarkdown::render(
      "inst/rmd/differential_expression.Rmd",
      params = list(
        path_config_json = "/beegfs/scratch/bruening_scratch/pklemm/2019-04-philipp-cers/nf-rnaseq/philipp_config.json",
        path_salmon_counts = "/beegfs/scratch/bruening_scratch/pklemm/2019-04-philipp-cers/nf-rnaseq/results/salmon/salmon_merged_gene_counts.csv"
      )
    )
}

rmarkdown_params <- list()
# Hack to be able to use infinite moon reader
if (params$dev == TRUE) {
  rmarkdown_params$path_config_json <- "/beegfs/scratch/bruening_scratch/pklemm/2019-04-philipp-cers/nf-rnaseq/philipp_config.json"
  rmarkdown_params$path_salmon_counts <- "/beegfs/scratch/bruening_scratch/pklemm/2019-04-philipp-cers/nf-rnaseq/results/salmon/salmon_merged_gene_counts.csv"
  rmarkdown_params$top_regulated_genes_for_plotting <- 10
} else {
  rmarkdown_params$path_config_json <- params$path_config_json
  rmarkdown_params$path_salmon_counts <- params$path_salmon_counts
  rmarkdown_params$top_regulated_genes_for_plotting <- params$top_regulated_genes_for_plotting
}

```

```{r setup, results = "hide"}
library(magrittr)
library(crosstalk)
library(plotly)
```

# DESeq2 Analysis

https://4va.github.io/biodatasci/r-rnaseq-airway.html

## Prepare DESeq2 Data Import

```{r load_data, message = FALSE}

settings <- jsonlite::read_json(rmarkdown_params$path_config_json)
count_data <- readr::read_csv(
  rmarkdown_params$path_salmon_counts,
  progress = FALSE
)

```

We need to prepare the `colData` table required by DESeq2 to determine the experiment design.

```{r prepare_data}

# Make empty tibble
col_data <- tibble::tibble()

# Iterate over group names and create treatment column
settings$groups %>%
  names() %>%
  purrr::walk(function(group_name) {
    coldata_group <- tibble::tibble(
      sample_name = settings$groups[group_name] %>% unlist(),
      treatment = group_name
    )
    # Attach rows to col_data tibble
    col_data <<- col_data %>%
      dplyr::bind_rows(coldata_group)
  })

# The order in coldata must be the same as in count_data
# We achieve this by joining based on the id in count_data
col_data <- tibble::tibble(
  id = count_data %>%
    # Get column names
    colnames() %>%
    # Remove "gene_id" column which is the first one
    .[2:length(.)]
) %>%
  dplyr::left_join(
    col_data,
    by = c("id" = "sample_name")
  )

```

```{r check_tables}
# Check if we have samples in count table that we do not have a treatment for
undefined_samples <- col_data %>%
  dplyr::filter(
    is.na(treatment)
  )

undefined_samples$id %>%
  purrr::walk(function(na_sample) {
    paste0("Removed sample ", na_sample, " because it was assigned to no treatment group in the JSON file") %>%
      warning()
    count_data <<- count_data %>%
      dplyr::select(-dplyr::contains(na_sample))
    }
  )

col_data %<>%
  dplyr::filter(!is.na(treatment))

# Convert counts to integer to comply with DEseq2 standards
count_data %<>%
  dplyr::mutate_if(is.double, round) %>%
  dplyr::mutate_if(is.double, as.integer)

```

Here are the top entries of the count table.

```{r print_top_entries}

count_data %>%
  head() %>%
  knitr::kable()

```

The experiment design table is defined as follows.

```{r print_design_table}

col_data %>%
  knitr::kable()

```

## Create DESeq2 Data and run DE

```{r create_deseq2_data}

deseq_data <- DESeq2::DESeqDataSetFromMatrix(
  countData = count_data %>% as.data.frame(),
  colData = col_data,
  design = ~treatment,
  tidy = TRUE
)

```

```{r run_differential_expression_analysis, cache = TRUE}

deseq_result <- DESeq2::DESeq(deseq_data)
deseq_result_frame <- DESeq2::results(
    deseq_result,
    tidy = TRUE
  ) %>%
  tibble::as_tibble()

```

There are `r deseq_result_frame %>% tibble::as_tibble() %>% dplyr::filter(padj <= 0.05) %>% nrow()` entries differentially expressed.

*If you click on one of the download buttons above the table, give your browser a minute to process the request.*

```{r print_de_entries}

deseq_result_frame %>%
  dplyr::filter(padj <= 0.05) %>%
  dplyr::mutate_if(is.double, function(x) { return(round(x, 4)) }) %>%
  DT::datatable(
    extensions = 'Buttons',
    options = list(
      dom = 'Bfrtip',
      buttons = c('copy', 'csv', 'excel', 'pdf', 'print')
    )
  )

```

# Plot results

Here we gather a couple of built-in plots of the final DESeq dataset.

```{r plot_results}

vsdata <- DESeq2::vst(deseq_result, blind = FALSE)
pca_plot <- DESeq2::plotPCA(vsdata, intgroup = "treatment")

plotly::ggplotly(pca_plot)

```

```{r ma_plot, cache = TRUE}
DESeq2::plotMA(
  DESeq2::lfcShrink(deseq_result, coef=2, type="ashr"),
  ylim=c(-2,2)
)
```

Heatmap of the sample-to-sample distances - https://bioconductor.org/packages/release/bioc/vignettes/DESeq2/inst/doc/DESeq2.html#differential-expression-analysis

```{r distance}

distance_matrix <- function(deseq_result) {
  vsdata <- DESeq2::vst(deseq_result, blind = FALSE)
  sampleDists <- dist(t(SummarizedExperiment::assay(vsdata)))
  sampleDistMatrix <- as.matrix(sampleDists)
  
  # rownames(sampleDistMatrix) <- paste(vsdata$condition, vsdata$type, sep="-")
  rownames(sampleDistMatrix) <- colnames(sampleDistMatrix)
  # colnames(sampleDistMatrix) <- NULL

  colors <- colorRampPalette( rev(RColorBrewer::brewer.pal(9, "Blues")) )(255)
  pheatmap::pheatmap(sampleDistMatrix,
    clustering_distance_rows = sampleDists,
    clustering_distance_cols = sampleDists,
    col = colors
  )
}

distance_matrix(deseq_result)
```

This can probably be moved to a separate markdown file to keep the footprint of the DESeq part low.

```{r get_deseq_counts, cache = TRUE}

deseq_counts <- deseq_result %>%
  DESeq2::counts(
    normalized = TRUE,
    replaced = FALSE
  ) %>%
  as.data.frame() %>%
  tibble::rownames_to_column("ensembl_id") %>%
  # Attach adjusted p-value
  dplyr::left_join(
    .,
    deseq_result_frame %>%
      dplyr::select(row, padj),
    by = c("ensembl_id" = "row")
  ) %>%
  # Order by adjusted p-value
  dplyr::arrange(padj) %>%
  rmyknife::attach_biomart(
    ensembl_version = 97,
    ensembl_id_var = "ensembl_id",
    attributes = "external_gene_name",
  ) %>%
  dplyr::rename(name = external_gene_name) %>%
  tibble::as_tibble()

```

```{r plot_genes, warning = FALSE}

deseq_counts_top_n <- deseq_counts %>%
  # Keep top n genes for plotting
  head(rmarkdown_params$top_regulated_genes_for_plotting) %>%
  # Remove padj
  dplyr::select(-padj, -ensembl_id) %>%
  tidyr::pivot_longer(-name, "sample") %>%
  # Attach treatment
  dplyr::left_join(
    col_data,
    by = c("sample" = "id")
  )

sd <- SharedData$new(deseq_counts_top_n)
filter_select(
  id = "name",
  label = "Select Gene",
  sharedData = sd,
  group = ~name,
  multiple = FALSE
  # selected does not work properly at the moment. See https://github.com/rstudio/crosstalk/pull/70 for details
  # selected = deseq_counts_top_n %>% 
  #   head(1) %>%
  #   dplyr::select(gene) %>%
  #   dplyr::pull()
  )

plot <- sd %>% ggplot2::ggplot(
    ggplot2::aes(
      x = treatment,
      y = value
    )
  ) +
  ggplot2::geom_point() +
  ggplot2::geom_boxplot()

plotly::ggplotly(plot)

```

https://genviz.org/module-04-expression/0004/02/01/DifferentialExpression/
